### ISSUE TITLE
Experiments details on the ablation study.

### ISSUE THREAD

**Issue by chenkang455** [wangst0181/pi-StepNFT#3]
@wangst0181 hi, thanks for the open-source of this nice work. Can you kindly provide some specific details on the ablation? (pi0? libero-spatial?)

**Comment by wangst0181** [comment]
Thanks for your interest! For the ablation study, we used Task 0 from the LIBERO Object suite. Since each LIBERO suite contains 10 tasks, running ablations on all tasks together would be very time-consuming. To make the ablation process faster and more practical for tuning hyperparameters, we selected one sub-task after SFT whose initial success rate was neither too high nor too low, making it a more suitable benchmark for comparing settings. This greatly speeds up the ablation cycle. On 8×4090 GPUs, we sample 32 trajectories per iteration and then train on them, and each ablation run takes about 2–3 hours.

**Comment by chenkang455** [comment]
@wangst0181 Get it! Thanks a lot! I am reproducing the experiment on libero_spatial with the pi0 sft model.However, I still have a small confusion regarding the paper. Based on my understanding of Diffusion-NFT, $v_{gt}$ is constructed by re-sampling noise. That is, given a generated action $a$, noise is added to obtain $a_t$, and then $v$ is calculated as $a - \text{noise}$.In the PI-StepNFT code, it seems the noise is not re-sampled from the generated action. Instead, it directly follows a specific step during the rollout. This raises an issue: we cannot obtain $v_{gt}$. I noticed the strategy here is to construct the loss by calculating x_next - mean_pos:
```
def _flow_mean(x_cur: torch.Tensor, velocity: torch.Tensor) -> torch.Tensor:
    x0_pred = x_cur - velocity * t_bc
    x1_pred = x_cur + velocity * (1 - t_bc)
    x0_weight = torch.ones_like(t_bc) - (t_bc - delta_bc)
    x1_weight = t_bc - delta_bc - sigma_i**2 * delta_bc / (2 * t_bc)
    return x0_pred * x0_weight + x1_pred * x1_weight

mean_pos = _flow_mean(x_t, v_pos)
mean_neg = _flow_mean(x_t, v_neg)
var = std_t_detached**2 + std_epsilon
E_pos = ((x_next - mean_pos) ** 2 / var).sum(dim=dims)
E_neg = ((x_next - mean_neg) ** 2 / var).sum(dim=dims)
delta_E = E_pos - E_neg

```
My confusion is: this essentially calculates the log-likelihood of the transition from $x$ to $x_{next}$, which makes it consistent with the PI-RL loss. I would like to ask the author if this understanding is correct.

**Comment by wangst0181** [comment]
Thanks for the careful question — this is a very good observation.

The key point is that our implementation, like Diffusion-NFT in practice, is **not** based on explicitly constructing a $v_{gt}$ and then applying an MSE directly in velocity space.

In the Diffusion-NFT code, the actual supervision is written in **$x$-space**: the model first uses the predicted velocity to reconstruct a predicted $x_0$, and then compares that reconstructed $x_0$ with the sampled $x_0$:

$x_0^{pred} = x_t - t \cdot v^{pred}.$

This is fully equivalent to supervising $v$ against the usual target derived from $x_t$ and $x_0$, just expressed in reconstruction form.

PI-StepNFT follows the same idea. We also write the loss in **state space**, but instead of using the final denoised target $x_0$, we use the **local denoising transition target**. In other words, we use the predicted velocity to reconstruct the predicted next denoising position $x_{t^-}^{pred}$, and then supervise it against the actual rollout successor $x_{t^-}$.

So your intuition is correct that this is related to modeling the transition from $x_t$ to $x_{t^-}$. That is intentional. The main difference from Diffusion-NFT is that we focus on a **local intermediate step**, rather than a full long-range path from initial noise to final $x_0$.

The reason is that, in our method, we replace ODE sampling with **SDE sampling during rollout** to enlarge the exploration space. Once stochasticity is injected during rollout, we need to recover the transition using the **actual rollout variance**, so that the supervision stays on the same denoising path that was actually sampled, instead of defining a new path by re-sampling noise from the final action. This is why the loss is written using $(x_{t^-} - mean_{pos}) / (x_{t^-} - mean_{neg})$ together with the rollout variance.

So to summarize:

1. We are **not** explicitly constructing a separate $v_{gt}$.
2. As in Diffusion-NFT, the supervision is written in **$x$-space**.
3. The main change is that we supervise a **step-wise local transition** $x_t \rightarrow x_{t^-}$, rather than the terminal reconstruction $x_0$.
4. This is exactly the motivation illustrated in Figure 1 of the paper: once exploration is widened by SDE, we need **finer step-wise supervision** instead of only supervising the final denoised result.

I hope this helps clarify the implementation difference.

**Comment by chenkang455** [comment]
@wangst0181 This modification is interesting, and thanks a lot for your detailed response!
